modules/consensus/database.go

```diff
@@ -56,417 +86,254 @@ func openDB(filename string) (*setDB, error) {
+	// consensus set hash being calculated
+	for _, sfod := range cs.blockRoot.SiafundOutputDiffs {
+		commitSiafundOutputDiff(tx, sfod, modules.DiffApply)
 	}
-	k := encoding.Marshal(key)
-	item := b.Get(k)
-	if item == nil {
-		return nil, errNilItem
-	}
-	return item, nil
-}
 
-// forEach iterates through a bucket, calling the supplied closure on each
-// element.
-func forEach(tx *bolt.Tx, bucket []byte, fn func(k, v []byte) error) error {
-	b := tx.Bucket(bucket)
-	if build.DEBUG && b == nil {
-		panic(errNilBucket)
-	}
-	return b.ForEach(fn)
-}
-
-// pushPath adds a block to the BlockPath at current height + 1.
-func pushPath(tx *bolt.Tx, bid types.BlockID) error {
-	b := tx.Bucket(BlockPath)
-	key := encoding.EncUint64(uint64(b.Stats().KeyN))
-	value := encoding.Marshal(bid)
-	return b.Put(key, value)
-}
-
-// getItem is a generic function to insert an item into the set database
-//
-// DEPRECATED
-func (db *setDB) getItem(bucket []byte, key interface{}) (item []byte, err error) {
-	k := encoding.Marshal(key)
-	err = db.View(func(tx *bolt.Tx) error {
-		b := tx.Bucket([]byte(bucket))
-		// Sanity check to make sure the bucket exists.
-		if build.DEBUG && b == nil {
-			panic(errNilBucket)
-		}
-		item = b.Get(k)
-		// Sanity check to make sure the item requested exists
-		if item == nil {
-			return errNilItem
-		}
-		return nil
+	// Add the miner payout from the genesis block to the delayed siacoin
+	// outputs - unspendable, as the unlock hash is blank.
+	createDSCOBucket(tx, types.MaturityDelay)
+	addDSCO(tx, types.MaturityDelay, cs.blockRoot.Block.MinerPayoutID(0), types.SiacoinOutput{
+		Value:      types.CalculateCoinbase(0),
+		UnlockHash: types.UnlockHash{},
 	})
-	return item, err
-}
 
-// rmItem removes an item from a bucket
-func (db *setDB) rmItem(bucket []byte, key interface{}) error {
-	k := encoding.Marshal(key)
-	return db.Update(func(tx *bolt.Tx) error {
-		b := tx.Bucket(bucket)
-		if build.DEBUG {
-			// Sanity check to make sure the bucket exists.
-			if b == nil {
-				panic(errNilBucket)
-			}
-			// Sanity check to make sure you are deleting an item that exists
-			item := b.Get(k)
-			if item == nil {
-				panic(errNilItem)
-			}
-		}
-		return b.Delete(k)
-	})
+	// Add the genesis block to the block strucutres - checksum must be taken
+	// after pushing the genesis block into the path.
+	pushPath(tx, cs.blockRoot.Block.ID())
+	if build.DEBUG {
+		cs.blockRoot.ConsensusChecksum = consensusChecksum(tx)
+	}
+	addBlockMap(tx, &cs.blockRoot)
+	return nil
 }
 
-// inBucket checks if an item with the given key is in the bucket
-func (db *setDB) inBucket(bucket []byte, key interface{}) bool {
-	exists, err := db.Exists(bucket, encoding.Marshal(key))
+// blockHeight returns the height of the blockchain.
+func blockHeight(tx *bolt.Tx) types.BlockHeight {
+	var height int
+	bh := tx.Bucket(BlockHeight)
+	err := encoding.Unmarshal(bh.Get(BlockHeight), &height)
```

Comment:
I think because it starts at -1?

Actually, I was pretty sure I made it a `types.BlockHeight`, but I guess that was in a commit that I had to roll back. (I ended up throwing away 100+ lines of changes more than a dozen times because the unit tests starting failing - lots of stuff got implemented a few times)